### PR TITLE
[FLINK-23063][python][table] Remove #connect

### PR DESCRIPTION
**Pull request by Airblader**
## What is the purpose of the change

This removes the deprecated `#connect` API and updates the documentation. The removal of the old descriptors has been split into a new sub-task.

## Brief change log

* Remove `#connect` from the entire stack.
* Update all documentation using `#connect` to use either the new `TableDescriptor`, or SQL DDL directly.

## Verifying this change

This change is already covered by existing tests.

## Does this pull request potentially affect one of the following parts:

  - Dependencies (does it add or upgrade a dependency): no
  - The public API, i.e., is any changed class annotated with `@Public(Evolving)`: yes
  - The serializers: no
  - The runtime per-record code paths (performance sensitive): no
  - Anything that affects deployment or recovery: JobManager (and its components), Checkpointing, Kubernetes/Yarn, ZooKeeper: no
  - The S3 file system connector: no

## Documentation

  - Does this pull request introduce a new feature? no
  - If yes, how is the feature documented? not applicable

### CODE REVIEW

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/TableScanTest.scala [hunk 1]
```diff
@@ -210,10 +210,13 @@ class TableScanTest extends TableTestBase {
 
   @Test
   def testTableApiScanWithTemporaryTable(): Unit = {
-    util.tableEnv.connect(new FileSystem().path(tempFolder.newFile.getPath))
-        .withFormat(new OldCsv())
-        .withSchema(new Schema().field("word", DataTypes.STRING()))
-        .createTemporaryTable("t1")
+    util.tableEnv.createTemporaryTable("t1", TableDescriptor.forConnector("datagen")
```
Comment (Airblader) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/TableScanTest.scala:213]: (The actual filesystem connector isn't available here, but we don't really need it anyway)

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogITCase.java [hunk 2]
```diff
@@ -157,37 +155,27 @@ public void testCsvTableViaAPI() throws Exception {
...
-                        .withFormat(format)
-                        .inAppendMode()
-                        .withComment("Comment.")
-                        .build();
+        final Map<String, String> sourceOptions = new HashMap<>();
+        sourceOptions.put("connector.type", "filesystem");
+        sourceOptions.put("connector.path", getClass().getResource("/csv/test.csv").getPath());
+        sourceOptions.put("format.type", "csv");
+
+        CatalogTable source = new CatalogTableImpl(schema, sourceOptions, "Comment.");
```
Comment (Airblader) [inline flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogITCase.java:169]: (I know this is deprecated, but it seems that the Hive implementation is still very much expecting this old stack and I didn't want to touch this here)

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/TableScanTest.scala [hunk 3]
```diff
@@ -210,10 +210,13 @@ class TableScanTest extends TableTestBase {
...
+      .schema(api.Schema.newBuilder()
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/plan/batch/sql/TableScanTest.scala:214]: remove `api.` prefix

File: flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/MemoryTableSourceSinkUtil.scala [hunk 4]
```diff
@@ -54,27 +56,30 @@ object MemoryTableSourceSinkUtil {
...
       tableName: String): Unit = {
-    tEnv.connect(new CustomConnectorDescriptor("LegacyUnsafeMemoryAppendTable", 1, false))
-      .withSchema(new Schema().schema(schema))
-      .createTemporaryTable(tableName)
+    tEnv.createTemporaryTable(tableName,
+      TableDescriptor.forConnector("LegacyUnsafeMemoryAppendTable")
+        .schema(schema.toSchema)
+        .build())
+
+    val sink = new UnsafeMemoryAppendTableSink
```
Comment (twalthr) [inline flink-table/flink-table-planner/src/test/scala/org/apache/flink/table/planner/utils/MemoryTableSourceSinkUtil.scala:72]: why are we registering two tables now?

File: flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogITCase.java (hunk 2 above)
Comment (twalthr) [reply flink-connectors/flink-connector-hive/src/test/java/org/apache/flink/table/catalog/hive/HiveCatalogITCase.java:169]: Makes sense.

**Review by twalthr — COMMENTED** [review]
Awesome work @Airblader. This is a big improvement for reducing confusion. I just had a minor comment.

**Review by twalthr — APPROVED** [review]
LGTM

**Pull request closed without merging**